### DISCUSSION TITLE
How to sync the state data saved in localstorage across all tabs using `Ngxs` state management?

### DISCUSSION THREAD

**Discussion by mahmudabir** [ngxs/store#1893] [Q&A] [answered]
I am saving all data of the state after converting it to base64 format using `Ngxs` state management library. I can get all data across all tabs. But if I change a data in one tab it does not sync in the other tabs. To sync the data I have to refresh the tab. Is there any way to sync full state in all tabs if the state is changed in one tab?

Here is my code of `AppModule`:

```
@NgModule({
  declarations: [
    AppComponent,
    ReadComponent,
    CreateComponent
  ],
  imports: [
    BrowserModule,
    AppRoutingModule,
    NgxsModule.forRoot([
      TutorialState
    ]),
    NgxsReduxDevtoolsPluginModule.forRoot(),
    NgxsLoggerPluginModule.forRoot(),
    NgxsStoragePluginModule.forRoot({
      storage: StorageOption.LocalStorage,
      serialize: (state) => btoa(JSON.stringify(state)),
      deserialize: (state) => JSON.parse(atob(state))
  ],
  providers: [],
  bootstrap: [AppComponent]
})
export class AppModule {
}
```

**Comment by Andreas-Hjortland** [comment]
I have actually created a plugin which does this by virtue of messaging between the tabs instead of trying to save it to a common store since I don't think ngxs is supposed to dynamically load the state from the store if it is changed (i might be wrong about it though)

I have created a demo for it [here](https://andreas-hjortland.github.io/ngxs-message-plugin/), and the source code is [here](https://github.com/Andreas-Hjortland/ngxs-message-plugin). It is on npm under the name `ngxs-message-plugin`

**Reply by mahmudabir** [reply]
I saw your demo. It does not sync state across other tabs.

**Reply by Andreas-Hjortland** [reply]
Hm... This is based on a master / client type of relationship between the windows, so it will only sync between between tabs that are opened from the main window and not completely separate tabs. It's probably possible to adapt this plugin to handle that use case as well, but it wouldn't be trivial

**Reply by mahmudabir** [reply]
Thank you, Can you suggest me a way to achieve my concern?

**Reply by Andreas-Hjortland** [reply]
You could perhaps do a combination of what I did in my plugin with the storage plugin... Save the state to local storage using the storage plugin, followed by broadcasting a message to every tab after the state is saved so that you could reload the state from every other tab, although I haven't tried it so the solution is copmletely theoretical

**Reply by mahmudabir** [reply]
Using the storage plugin saving the state into local Storage & get them in other tabs works automatically. But I purpose is to achieve the feature reactively, i.e., If I change a data into a tab it must reflect to other tabs as well. So how can I set the data of state to the store manually?

**Reply by Andreas-Hjortland** [reply]
Yeah, thats what I am not completely sure about... What I was thinking was to use the [BroadcastChannel API](https://developer.mozilla.org/en-US/docs/Web/API/Broadcast_Channel_API) to notify the other tabs whenever the state changes that it has changed and we have to reload the state from localStorage. I am not 100% sure how to do that though, and unfortunately I am not able to figure it out at the moment either. I hope it's enough of a starting point so that you might be able to find a solution though

**Reply by mahmudabir** [reply]
ok, that is one thing to implement. But can you please tell me how to set data in the store. I am able to get all data from localstorage reactively. But don't know how to set the data in a store. I searched for it. but could not get any solution.

**Comment by kuncevic** [comment, marked as answer]
@mahmudabir also check this discussion  https://github.com/ngxs/store/issues/793 it reference a blogpost that may help

**Reply by mahmudabir** [reply]
Thank you. I have solved this issue using fromEvent by Rxjs.

**Reply by dinbtechit** [reply]
@mahmudabir  - It is causing an  infinite loop situtation

Tab 1 updates the state, which in turn updates the local storage. This triggers a new event in Tab 2, causing the state to be updated and the storage to be updated as well. The process continues as new events are sent back and forth between the tabs. 

How did you get around this?
